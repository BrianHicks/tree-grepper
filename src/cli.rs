use crate::extractor::Extractor;
use crate::extractor_chooser::ExtractorChooser;
use crate::language::Language;
use anyhow::{bail, Context, Error, Result};
use clap::{crate_authors, crate_version, App, Arg, ArgMatches};
use itertools::Itertools;
use std::collections::HashMap;
use std::path::PathBuf;
use std::str::FromStr;

pub enum Invocation {
    DoQuery(QueryOpts),
    ShowLanguages,
}

#[derive(Debug)]
pub struct QueryOpts {
    pub extractors: Vec<Extractor>,
    pub paths: Vec<PathBuf>,
    pub git_ignore: bool,
    pub format: QueryFormat,
    pub sort: bool,
}

impl QueryOpts {
    pub fn extractor_chooser(&self) -> Result<ExtractorChooser> {
        ExtractorChooser::from_extractors(&self.extractors)
    }
}

impl Invocation {
    pub fn from_args(args: Vec<String>) -> Result<Self> {
        // I'm not super happy with this! I would love for LANGUAGE and QUERY to
        // be taken positionally when there is just one so we don't always have
        // to specify `-q`. However, I also want to get working on the rest of
        // the program so I'm dropping the requirement for now by making `-q`
        // required. I think that's an OK tradeoff until I can figure something
        // else better because it'll be backwards compatible with the scheme
        // I outlined above.
        //
        // Check
        // https://users.rust-lang.org/t/grep-like-argument-parsing-with-clap/63392
        // for where I asked about this in public.
        let matches = App::new("tree-grepper")
            .version(crate_version!())
            .author(crate_authors!())
            .arg(
                Arg::new("additional-query")
                    .short('q')
                    .long("query")
                    .help("a language and query to perform")
                    .long_help(
                        "a language and query to perform (at least one is required.) See https://tree-sitter.github.io for information on writing queries. Run tree-grepper --languages for a list of languages.",
                    )
                    .number_of_values(2)
                    .value_names(&["LANGUAGE", "QUERY"])
                    .required_unless_present("languages")
                    .multiple_values(true)
            )
            .arg(
                Arg::new("no-gitignore")
                    .long("no-gitignore")
                    .help("don't use git's ignore and exclude files to filter files")
            )
            .arg(
                Arg::new("PATHS")
                    .default_value(".")
                    .help("places to search for matches")
                    .multiple_values(true)
            )
            .arg(
                Arg::new("FORMAT")
                .long("format")
                .short('f')
                .possible_values(&["lines", "json", "json-lines", "pretty-json"])
                .default_value("lines")
                .help("what format should we output lines in?")
            )
            .arg(
                Arg::new("sort")
                .long("sort")
                .help("sort matches stably")
                .long_help("sort matches stably. If this is not specified, output ordering will vary because due to parallelism. Caution: this adds a worst-case `O(n * log(n))` overhead, where `n` is the number of files matched. Avoid it if possible if you care about performance.")
            )
            .arg(
                Arg::new("languages")
                .long("languages")
                .help("print the language names tree-grepper knows about")
            )
            .try_get_matches_from(args)
            .context("could not parse args")?;

        if matches.is_present("languages") {
            Ok(Self::ShowLanguages)
        } else {
            Ok(Self::DoQuery(QueryOpts {
                extractors: Self::extractors(&matches)?,
                paths: Self::paths(&matches)?,
                git_ignore: !matches.is_present("no-gitignore"),
                format: QueryFormat::from_str(
                    matches.value_of("FORMAT").context("format not provided")?,
                )
                .context("could not set format")?,
                sort: matches.is_present("sort"),
            }))
        }
    }

    fn extractors(matches: &ArgMatches) -> Result<Vec<Extractor>> {
        let values = match matches.values_of("additional-query") {
            Some(values) => values,
            None => bail!("queries were required but not provided. This indicates an internal error and you should report it!"),
        };

        // the most common case is going to be one query, so let's allocate
        // that immediately...
        let mut query_strings: HashMap<Language, String> = HashMap::with_capacity(1);

        // If you have two tree-sitter queries `(one)` and `(two)`, you can
        // join them together in a single string like `(one)(two)`. In that
        // case, the resulting query will act like an OR and match any of the
        // queries inside. Doing this automatically gives us an advantage:
        // for however many queries we get on the command line, we will only
        // ever have to run one per file, since we can combine them and you
        // can't specify queries across multiple languages! Nobody should ever
        // notice, except that they won't see as much of a slowdown for adding
        // new queries to an invocation as they might expect. (Well, hopefully!)
        for (raw_lang, raw_query) in values.tuples() {
            let lang = Language::from_str(raw_lang).context("could not parse language")?;

            let mut query_out = String::from(raw_query);

            let temp_query = lang
                .parse_query(raw_query)
                .context("could not parse query")?;

            if temp_query.capture_names().is_empty() {
                query_out.push_str("@query");
            }

            if let Some(existing) = query_strings.get_mut(&lang) {
                existing.push_str(&query_out);
            } else {
                query_strings.insert(lang, query_out);
            }
        }

        let mut out = Vec::with_capacity(query_strings.len());
        for (lang, raw_query) in query_strings {
            let query = lang
                .parse_query(&raw_query)
                .context("could not parse combined query")?;

            out.push(Extractor::new(lang, query))
        }

        Ok(out)
    }

    fn paths(matches: &ArgMatches) -> Result<Vec<PathBuf>> {
        match matches.values_of("PATHS") {
            Some(values) =>
                values
                    .map(|raw_path| PathBuf::from_str(raw_path).with_context(|| format!("could not parse a path from {}", raw_path)))
                    .collect(),

            None => bail!("at least one path was required but not provided. This indicates an internal errors and you should report it!"),
        }
    }
}

#[derive(Debug)]
pub enum QueryFormat {
    Lines,
    Json,
    JsonLines,
    PrettyJson,
}

impl FromStr for QueryFormat {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> {
        match s {
            "lines" => Ok(QueryFormat::Lines),
            "json" => Ok(QueryFormat::Json),
            "json-lines" => Ok(QueryFormat::JsonLines),
            "pretty-json" => Ok(QueryFormat::PrettyJson),
            _ => bail!("unknown format. See --help for valid formats."),
        }
    }
}
