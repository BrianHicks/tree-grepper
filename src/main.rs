mod cli;
mod extractor;
mod extractor_chooser;
mod language;

use anyhow::{bail, Context, Result};
use cli::{Format, Opts};
use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
use std::env;
use std::io::{self, Write};
use tree_sitter::Parser;

#[global_allocator]
static ALLOCATOR: bump_alloc::BumpAlloc = bump_alloc::BumpAlloc::new();

fn main() {
    if let Err(error) = try_main(env::args().collect(), &io::stdout()) {
        // Clap errors (--help or misuse) are already well-formatted, so we
        // don't have to do any additional work.
        if let Some(clap_error) = error.downcast_ref::<clap::Error>() {
            eprintln!("{}", clap_error);
        } else {
            eprintln!("{:?}", error);
        }
        std::process::exit(1);
    }
}

fn try_main(args: Vec<String>, mut out: impl Write) -> Result<()> {
    let opts = Opts::from_args(args)
        .context("couldn't get a valid configuration from the command-line options")?;

    // You might think "why not use ParallelBridge here?" Well, the quick answer
    // is that I benchmarked it and having things separated here and handling
    // their own errors actually speeds up this part of the code by like 20%!
    let items: Vec<ignore::DirEntry> = build_walker(&opts)
        .context("couldn't build a filesystem walker")?
        .collect::<Result<Vec<ignore::DirEntry>, ignore::Error>>()
        .context("had a problem while walking the filesystem")?;

    let chooser = opts
        .extractor_chooser()
        .context("couldn't construct a filetype matcher")?;

    let mut extracted_files = items
        .par_iter()
        .filter_map(|entry| {
            chooser
                .extractor_for(entry)
                .map(|extractor| (entry, extractor))
        })
        .map_init(Parser::new, |parser, (entry, extractor)| {
            extractor.extract_from_file(entry.path(), parser)
        })
        .filter_map(|result_containing_option| match result_containing_option {
            Ok(None) => None,
            Ok(Some(extraction)) => Some(Ok(extraction)),
            Err(err) => Some(Err(err)),
        })
        .collect::<Result<Vec<extractor::ExtractedFile>>>()
        .context("couldn't extract matches from files")?;

    if opts.sort {
        extracted_files.sort()
    }

    match opts.format {
        Format::Lines => {
            for extracted_file in extracted_files {
                write!(out, "{}", extracted_file).context("could not write lines")?;
            }
        }

        Format::Json => {
            serde_json::to_writer(out, &extracted_files).context("could not write JSON output")?;
        }

        Format::PrettyJson => {
            serde_json::to_writer_pretty(out, &extracted_files)
                .context("could not write JSON output")?;
        }
    }

    Ok(())
}

fn build_walker(opts: &Opts) -> Result<ignore::Walk> {
    let mut builder = match opts.paths.split_first() {
        Some((first, rest)) => {
            let mut builder = ignore::WalkBuilder::new(first);
            for path in rest {
                builder.add(path);
            }

            builder
        }
        None => bail!("I need at least one file or directory to walk!"),
    };

    Ok(builder
        .git_ignore(opts.git_ignore)
        .git_exclude(opts.git_ignore)
        .git_global(opts.git_ignore)
        .build())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn call(args: &[&str]) -> String {
        let mut bytes = Vec::new();
        try_main(
            args.iter().map(|s| s.to_string()).collect(),
            Box::new(&mut bytes),
        )
        .unwrap();

        String::from_utf8(bytes).unwrap()
    }

    #[test]
    fn lines_output() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "elm",
            "(import_clause)",
            "-f",
            "lines",
            "--sort",
            "vendor/tree-sitter-elm",
        ]))
    }

    #[test]
    fn json_output() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "elm",
            "(import_clause)",
            "-f",
            "json",
            "--sort",
            "vendor/tree-sitter-elm",
        ]))
    }

    #[test]
    fn pretty_json_output() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "elm",
            "(import_clause)",
            "--format=pretty-json",
            "--sort",
            "vendor/tree-sitter-elm",
        ]))
    }

    // All languages should have a test that just spits out their entire node
    // tree. We use this to know about changes in the vendored parsers!

    #[test]
    fn all_elm() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "elm",
            "(_)",
            "--format=pretty-json",
            "--sort",
            "vendor/tree-sitter-elm",
        ]))
    }

    #[test]
    fn all_haskell() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "haskell",
            "(_)",
            "--format=pretty-json",
            "--sort",
            "vendor/tree-sitter-haskell",
        ]))
    }

    #[test]
    fn all_javascript() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "javascript",
            "(_)",
            "--format=pretty-json",
            "--sort",
            // note that this doesn't include the entire vendor
            // directory. tree-sitter-javascript vendors a couple of libraries
            // to test things and it makes this test run unacceptably long. I
            // think the slowdown is due to the diffing step; the tree-grepper
            // code completes in a reasonable amount of time.
            "vendor/tree-sitter-javascript/test",
        ]))
    }

    #[test]
    fn all_ruby() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "ruby",
            "(_)",
            "--format=pretty-json",
            "--sort",
            "vendor/tree-sitter-ruby",
        ]))
    }

    #[test]
    fn all_rust() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "rust",
            "(_)",
            "--format=pretty-json",
            "--sort",
            "vendor/tree-sitter-rust",
        ]))
    }

    #[test]
    fn all_typescript() {
        insta::assert_snapshot!(call(&[
            "tree-grepper",
            "-q",
            "typescript",
            "(_)",
            "--format=pretty-json",
            "--sort",
            // similar to JavaScript, there is one particular test file in this
            // grammar that's *huge*. It seems to be a comprehensive listing of
            // all the typescript syntax, maybe? Regardless, it makes this test
            // unacceptably slow, so we just look at one particular file. If
            // we see uncaught regressions in this function, we probably will
            // make our own test file with the things we care about.
            "vendor/tree-sitter-typescript/typescript/test.ts",
        ]))
    }
}
