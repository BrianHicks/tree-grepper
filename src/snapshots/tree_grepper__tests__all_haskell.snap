---
source: src/main.rs
expression: "call(&[\"tree-grepper\", \"-q\", \"haskell\", \"(_)\", \"--format=pretty-json\",\n       \"--sort\", \"vendor/tree-sitter-haskell\"])"

---
[
  {
    "file": "vendor/tree-sitter-haskell/test/edit/Test.hs",
    "matches": [
      {
        "kind": "haskell",
        "name": "query",
        "text": "module A where\n\ndata Foo a =\n  Foo\n  deriving (Eq)\n  deriving (Show)\n",
        "start": {
          "row": 0,
          "column": 0
        },
        "end": {
          "row": 6,
          "column": 0
        }
      },
      {
        "kind": "module",
        "name": "query",
        "text": "module",
        "start": {
          "row": 0,
          "column": 0
        },
        "end": {
          "row": 0,
          "column": 6
        }
      },
      {
        "kind": "module",
        "name": "query",
        "text": "A",
        "start": {
          "row": 0,
          "column": 7
        },
        "end": {
          "row": 0,
          "column": 8
        }
      },
      {
        "kind": "where",
        "name": "query",
        "text": "where",
        "start": {
          "row": 0,
          "column": 9
        },
        "end": {
          "row": 0,
          "column": 14
        }
      },
      {
        "kind": "adt",
        "name": "query",
        "text": "data Foo a =\n  Foo\n  deriving (Eq)\n  deriving (Show)",
        "start": {
          "row": 2,
          "column": 0
        },
        "end": {
          "row": 5,
          "column": 17
        }
      },
      {
        "kind": "data",
        "name": "query",
        "text": "data",
        "start": {
          "row": 2,
          "column": 0
        },
        "end": {
          "row": 2,
          "column": 4
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "Foo",
        "start": {
          "row": 2,
          "column": 5
        },
        "end": {
          "row": 2,
          "column": 8
        }
      },
      {
        "kind": "type_variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 2,
          "column": 9
        },
        "end": {
          "row": 2,
          "column": 10
        }
      },
      {
        "kind": "=",
        "name": "query",
        "text": "=",
        "start": {
          "row": 2,
          "column": 11
        },
        "end": {
          "row": 2,
          "column": 12
        }
      },
      {
        "kind": "constructors",
        "name": "query",
        "text": "Foo",
        "start": {
          "row": 3,
          "column": 2
        },
        "end": {
          "row": 3,
          "column": 5
        }
      },
      {
        "kind": "data_constructor",
        "name": "query",
        "text": "Foo",
        "start": {
          "row": 3,
          "column": 2
        },
        "end": {
          "row": 3,
          "column": 5
        }
      },
      {
        "kind": "constructor",
        "name": "query",
        "text": "Foo",
        "start": {
          "row": 3,
          "column": 2
        },
        "end": {
          "row": 3,
          "column": 5
        }
      },
      {
        "kind": "deriving",
        "name": "query",
        "text": "deriving (Eq)",
        "start": {
          "row": 4,
          "column": 2
        },
        "end": {
          "row": 4,
          "column": 15
        }
      },
      {
        "kind": "deriving",
        "name": "query",
        "text": "deriving",
        "start": {
          "row": 4,
          "column": 2
        },
        "end": {
          "row": 4,
          "column": 10
        }
      },
      {
        "kind": "(",
        "name": "query",
        "text": "(",
        "start": {
          "row": 4,
          "column": 11
        },
        "end": {
          "row": 4,
          "column": 12
        }
      },
      {
        "kind": "constraint",
        "name": "query",
        "text": "Eq",
        "start": {
          "row": 4,
          "column": 12
        },
        "end": {
          "row": 4,
          "column": 14
        }
      },
      {
        "kind": "class_name",
        "name": "query",
        "text": "Eq",
        "start": {
          "row": 4,
          "column": 12
        },
        "end": {
          "row": 4,
          "column": 14
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "Eq",
        "start": {
          "row": 4,
          "column": 12
        },
        "end": {
          "row": 4,
          "column": 14
        }
      },
      {
        "kind": ")",
        "name": "query",
        "text": ")",
        "start": {
          "row": 4,
          "column": 14
        },
        "end": {
          "row": 4,
          "column": 15
        }
      },
      {
        "kind": "deriving",
        "name": "query",
        "text": "deriving (Show)",
        "start": {
          "row": 5,
          "column": 2
        },
        "end": {
          "row": 5,
          "column": 17
        }
      },
      {
        "kind": "deriving",
        "name": "query",
        "text": "deriving",
        "start": {
          "row": 5,
          "column": 2
        },
        "end": {
          "row": 5,
          "column": 10
        }
      },
      {
        "kind": "(",
        "name": "query",
        "text": "(",
        "start": {
          "row": 5,
          "column": 11
        },
        "end": {
          "row": 5,
          "column": 12
        }
      },
      {
        "kind": "constraint",
        "name": "query",
        "text": "Show",
        "start": {
          "row": 5,
          "column": 12
        },
        "end": {
          "row": 5,
          "column": 16
        }
      },
      {
        "kind": "class_name",
        "name": "query",
        "text": "Show",
        "start": {
          "row": 5,
          "column": 12
        },
        "end": {
          "row": 5,
          "column": 16
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "Show",
        "start": {
          "row": 5,
          "column": 12
        },
        "end": {
          "row": 5,
          "column": 16
        }
      },
      {
        "kind": ")",
        "name": "query",
        "text": ")",
        "start": {
          "row": 5,
          "column": 16
        },
        "end": {
          "row": 5,
          "column": 17
        }
      }
    ]
  },
  {
    "file": "vendor/tree-sitter-haskell/test/highlight/Basic.hs",
    "matches": [
      {
        "kind": "haskell",
        "name": "query",
        "text": "module A where\n\ndata A a = A a\n\nclass C a where\n\na :: C a => A a -> A a\na (A a) =\n  A a\n",
        "start": {
          "row": 0,
          "column": 0
        },
        "end": {
          "row": 9,
          "column": 0
        }
      },
      {
        "kind": "module",
        "name": "query",
        "text": "module",
        "start": {
          "row": 0,
          "column": 0
        },
        "end": {
          "row": 0,
          "column": 6
        }
      },
      {
        "kind": "module",
        "name": "query",
        "text": "A",
        "start": {
          "row": 0,
          "column": 7
        },
        "end": {
          "row": 0,
          "column": 8
        }
      },
      {
        "kind": "where",
        "name": "query",
        "text": "where",
        "start": {
          "row": 0,
          "column": 9
        },
        "end": {
          "row": 0,
          "column": 14
        }
      },
      {
        "kind": "adt",
        "name": "query",
        "text": "data A a = A a",
        "start": {
          "row": 2,
          "column": 0
        },
        "end": {
          "row": 2,
          "column": 14
        }
      },
      {
        "kind": "data",
        "name": "query",
        "text": "data",
        "start": {
          "row": 2,
          "column": 0
        },
        "end": {
          "row": 2,
          "column": 4
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "A",
        "start": {
          "row": 2,
          "column": 5
        },
        "end": {
          "row": 2,
          "column": 6
        }
      },
      {
        "kind": "type_variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 2,
          "column": 7
        },
        "end": {
          "row": 2,
          "column": 8
        }
      },
      {
        "kind": "=",
        "name": "query",
        "text": "=",
        "start": {
          "row": 2,
          "column": 9
        },
        "end": {
          "row": 2,
          "column": 10
        }
      },
      {
        "kind": "constructors",
        "name": "query",
        "text": "A a",
        "start": {
          "row": 2,
          "column": 11
        },
        "end": {
          "row": 2,
          "column": 14
        }
      },
      {
        "kind": "data_constructor",
        "name": "query",
        "text": "A a",
        "start": {
          "row": 2,
          "column": 11
        },
        "end": {
          "row": 2,
          "column": 14
        }
      },
      {
        "kind": "constructor",
        "name": "query",
        "text": "A",
        "start": {
          "row": 2,
          "column": 11
        },
        "end": {
          "row": 2,
          "column": 12
        }
      },
      {
        "kind": "type_name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 2,
          "column": 13
        },
        "end": {
          "row": 2,
          "column": 14
        }
      },
      {
        "kind": "type_variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 2,
          "column": 13
        },
        "end": {
          "row": 2,
          "column": 14
        }
      },
      {
        "kind": "class",
        "name": "query",
        "text": "class C a where",
        "start": {
          "row": 4,
          "column": 0
        },
        "end": {
          "row": 4,
          "column": 15
        }
      },
      {
        "kind": "class",
        "name": "query",
        "text": "class",
        "start": {
          "row": 4,
          "column": 0
        },
        "end": {
          "row": 4,
          "column": 5
        }
      },
      {
        "kind": "class_head",
        "name": "query",
        "text": "C a",
        "start": {
          "row": 4,
          "column": 6
        },
        "end": {
          "row": 4,
          "column": 9
        }
      },
      {
        "kind": "class_name",
        "name": "query",
        "text": "C",
        "start": {
          "row": 4,
          "column": 6
        },
        "end": {
          "row": 4,
          "column": 7
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "C",
        "start": {
          "row": 4,
          "column": 6
        },
        "end": {
          "row": 4,
          "column": 7
        }
      },
      {
        "kind": "type_name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 4,
          "column": 8
        },
        "end": {
          "row": 4,
          "column": 9
        }
      },
      {
        "kind": "type_variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 4,
          "column": 8
        },
        "end": {
          "row": 4,
          "column": 9
        }
      },
      {
        "kind": "class_body",
        "name": "query",
        "text": "where",
        "start": {
          "row": 4,
          "column": 10
        },
        "end": {
          "row": 4,
          "column": 15
        }
      },
      {
        "kind": "where",
        "name": "query",
        "text": "where",
        "start": {
          "row": 4,
          "column": 10
        },
        "end": {
          "row": 4,
          "column": 15
        }
      },
      {
        "kind": "signature",
        "name": "query",
        "text": "a :: C a => A a -> A a",
        "start": {
          "row": 6,
          "column": 0
        },
        "end": {
          "row": 6,
          "column": 22
        }
      },
      {
        "kind": "variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 6,
          "column": 0
        },
        "end": {
          "row": 6,
          "column": 1
        }
      },
      {
        "kind": "::",
        "name": "query",
        "text": "::",
        "start": {
          "row": 6,
          "column": 2
        },
        "end": {
          "row": 6,
          "column": 4
        }
      },
      {
        "kind": "context",
        "name": "query",
        "text": "C a => A a -> A a",
        "start": {
          "row": 6,
          "column": 5
        },
        "end": {
          "row": 6,
          "column": 22
        }
      },
      {
        "kind": "constraint",
        "name": "query",
        "text": "C a",
        "start": {
          "row": 6,
          "column": 5
        },
        "end": {
          "row": 6,
          "column": 8
        }
      },
      {
        "kind": "class_name",
        "name": "query",
        "text": "C",
        "start": {
          "row": 6,
          "column": 5
        },
        "end": {
          "row": 6,
          "column": 6
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "C",
        "start": {
          "row": 6,
          "column": 5
        },
        "end": {
          "row": 6,
          "column": 6
        }
      },
      {
        "kind": "type_name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 6,
          "column": 7
        },
        "end": {
          "row": 6,
          "column": 8
        }
      },
      {
        "kind": "type_variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 6,
          "column": 7
        },
        "end": {
          "row": 6,
          "column": 8
        }
      },
      {
        "kind": "=>",
        "name": "query",
        "text": "=>",
        "start": {
          "row": 6,
          "column": 9
        },
        "end": {
          "row": 6,
          "column": 11
        }
      },
      {
        "kind": "fun",
        "name": "query",
        "text": "A a -> A a",
        "start": {
          "row": 6,
          "column": 12
        },
        "end": {
          "row": 6,
          "column": 22
        }
      },
      {
        "kind": "type_apply",
        "name": "query",
        "text": "A a",
        "start": {
          "row": 6,
          "column": 12
        },
        "end": {
          "row": 6,
          "column": 15
        }
      },
      {
        "kind": "type_name",
        "name": "query",
        "text": "A",
        "start": {
          "row": 6,
          "column": 12
        },
        "end": {
          "row": 6,
          "column": 13
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "A",
        "start": {
          "row": 6,
          "column": 12
        },
        "end": {
          "row": 6,
          "column": 13
        }
      },
      {
        "kind": "type_name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 6,
          "column": 14
        },
        "end": {
          "row": 6,
          "column": 15
        }
      },
      {
        "kind": "type_variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 6,
          "column": 14
        },
        "end": {
          "row": 6,
          "column": 15
        }
      },
      {
        "kind": "->",
        "name": "query",
        "text": "->",
        "start": {
          "row": 6,
          "column": 16
        },
        "end": {
          "row": 6,
          "column": 18
        }
      },
      {
        "kind": "type_apply",
        "name": "query",
        "text": "A a",
        "start": {
          "row": 6,
          "column": 19
        },
        "end": {
          "row": 6,
          "column": 22
        }
      },
      {
        "kind": "type_name",
        "name": "query",
        "text": "A",
        "start": {
          "row": 6,
          "column": 19
        },
        "end": {
          "row": 6,
          "column": 20
        }
      },
      {
        "kind": "type",
        "name": "query",
        "text": "A",
        "start": {
          "row": 6,
          "column": 19
        },
        "end": {
          "row": 6,
          "column": 20
        }
      },
      {
        "kind": "type_name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 6,
          "column": 21
        },
        "end": {
          "row": 6,
          "column": 22
        }
      },
      {
        "kind": "type_variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 6,
          "column": 21
        },
        "end": {
          "row": 6,
          "column": 22
        }
      },
      {
        "kind": "function",
        "name": "query",
        "text": "a (A a) =\n  A a",
        "start": {
          "row": 7,
          "column": 0
        },
        "end": {
          "row": 8,
          "column": 5
        }
      },
      {
        "kind": "variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 7,
          "column": 0
        },
        "end": {
          "row": 7,
          "column": 1
        }
      },
      {
        "kind": "patterns",
        "name": "query",
        "text": "(A a)",
        "start": {
          "row": 7,
          "column": 2
        },
        "end": {
          "row": 7,
          "column": 7
        }
      },
      {
        "kind": "pat_parens",
        "name": "query",
        "text": "(A a)",
        "start": {
          "row": 7,
          "column": 2
        },
        "end": {
          "row": 7,
          "column": 7
        }
      },
      {
        "kind": "(",
        "name": "query",
        "text": "(",
        "start": {
          "row": 7,
          "column": 2
        },
        "end": {
          "row": 7,
          "column": 3
        }
      },
      {
        "kind": "pat_apply",
        "name": "query",
        "text": "A a",
        "start": {
          "row": 7,
          "column": 3
        },
        "end": {
          "row": 7,
          "column": 6
        }
      },
      {
        "kind": "pat_name",
        "name": "query",
        "text": "A",
        "start": {
          "row": 7,
          "column": 3
        },
        "end": {
          "row": 7,
          "column": 4
        }
      },
      {
        "kind": "constructor",
        "name": "query",
        "text": "A",
        "start": {
          "row": 7,
          "column": 3
        },
        "end": {
          "row": 7,
          "column": 4
        }
      },
      {
        "kind": "pat_name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 7,
          "column": 5
        },
        "end": {
          "row": 7,
          "column": 6
        }
      },
      {
        "kind": "variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 7,
          "column": 5
        },
        "end": {
          "row": 7,
          "column": 6
        }
      },
      {
        "kind": ")",
        "name": "query",
        "text": ")",
        "start": {
          "row": 7,
          "column": 6
        },
        "end": {
          "row": 7,
          "column": 7
        }
      },
      {
        "kind": "=",
        "name": "query",
        "text": "=",
        "start": {
          "row": 7,
          "column": 8
        },
        "end": {
          "row": 7,
          "column": 9
        }
      },
      {
        "kind": "exp_apply",
        "name": "query",
        "text": "A a",
        "start": {
          "row": 8,
          "column": 2
        },
        "end": {
          "row": 8,
          "column": 5
        }
      },
      {
        "kind": "exp_name",
        "name": "query",
        "text": "A",
        "start": {
          "row": 8,
          "column": 2
        },
        "end": {
          "row": 8,
          "column": 3
        }
      },
      {
        "kind": "constructor",
        "name": "query",
        "text": "A",
        "start": {
          "row": 8,
          "column": 2
        },
        "end": {
          "row": 8,
          "column": 3
        }
      },
      {
        "kind": "exp_name",
        "name": "query",
        "text": "a",
        "start": {
          "row": 8,
          "column": 4
        },
        "end": {
          "row": 8,
          "column": 5
        }
      },
      {
        "kind": "variable",
        "name": "query",
        "text": "a",
        "start": {
          "row": 8,
          "column": 4
        },
        "end": {
          "row": 8,
          "column": 5
        }
      }
    ]
  }
]
